src/Network.cpp

```diff
@@ -495,20 +510,47 @@ T relative_difference(const T a, const T b) {
     return fabs(fa - fb) / std::min(fa, fb);
 }
 
-void compare_net_outputs(std::vector<float>& data,
-                         std::vector<float>& ref) {
-    // We accept an error up to 5%, but output values
-    // smaller than 1/1000th are "rounded up" for the comparison.
-    constexpr auto relative_error = 5e-2f;
-    for (auto idx = size_t{0}; idx < data.size(); ++idx) {
-        const auto err = relative_difference(data[idx], ref[idx]);
+void Network::compare_net_outputs(Netresult& data,
+                                  Netresult& ref) {
+    // We accept an error up to 20%, but output values
+    // smaller than 1/361th are "rounded up" for the comparison.
+    constexpr auto relative_error = 2e-1f;
+
+    // assert-fail when we hit 3 failures out of last 10 checks
+    constexpr auto max_failures = 3;
+    constexpr auto last_failure_window = 10;
+
+    auto selfcheck_fail = false;
+    for (auto idx = size_t{0}; idx < data.policy.size(); ++idx) {
+        const auto err = relative_difference(data.policy[idx], ref.policy[idx]);
         if (err > relative_error) {
-            printf("Error in OpenCL calculation: expected %f got %f "
-                   "(error=%f%%)\n", ref[idx], data[idx], err * 100.0);
-            printf("Update your GPU drivers or reduce the amount of games "
+            selfcheck_fail = true;
+            break;
+        }
+    }
+    const auto err_pass = relative_difference(data.policy_pass, ref.policy_pass);
+    const auto err_winrate = relative_difference(data.winrate, ref.winrate);
+    if (err_pass > relative_error) {
+        selfcheck_fail = true;
+    }
+    if (err_winrate > relative_error) {
+        selfcheck_fail = true;
+    }
+
+    LOCK(m_selfcheck_mutex, selfcheck_lock);
```

Comment:
I think you might be able to get rid of this lock by making m_selfcheck_fails a bitfield and CAS-ing it. (Optional)

Reply:
Thought about quickly, but it seems to be more trouble than necessary - we need to implement an array of bitfields and then implement a circular buffer to track last N pass/fails.  This lock is unlikely to be performance-critical anyway (only happens one out of 2000 evals).

Any better idea is welcome but I can't figure out how to make it simple :)

Reply:
>only happens one out of 2000 evals

That's a good point and makes my idea moot.

For reference:
You don't need an array or circular buffer. You just add the latest result like this ```bitfield = (bitfield << 1 | result) & 0xA;``` i.e. just relying on the shift and mask. Checking failure then is popcount(bitfield) > 2. Because the bitfield can be an int32, it can be atomically CASed.